src/Ledger/TokenAlgebra/ValueSet.lagda

```diff
@@ -0,0 +1,221 @@
+  ≋-isEquivalence : IsEquivalence {0ℓ} _≋_
+  ≋-isEquivalence = record { refl = ≡.refl ; sym = λ x → ≡.sym x ; trans = λ x y → ≡.trans x y }
+
+  ⊕-cong : Algebra.Congruent₂ _≋_ _⊕_
+  ⊕-cong {x}{y}{u}{v} x≋y u≋v {a} = begin
+    lookup (x ⊕ u) a         ≡⟨ ⊕-lemma x u ⟩
+    lookup x a + lookup u a  ≡⟨ cong (λ w → w + lookup u a) x≋y ⟩
+    lookup y a + lookup u a  ≡⟨ cong (λ w → lookup y a + w) u≋v ⟩
+    lookup y a + lookup v a  ≡⟨ ≡.sym (⊕-lemma y v) ⟩
+    lookup (y ⊕ v) a         ∎
+
+  ⊕-comm : Algebra.Commutative _≋_ _⊕_
+  ⊕-comm u v {aa} = begin
+    lookup (u ⊕ v) aa          ≡⟨ ⊕-lemma u v ⟩
+    lookup u aa + lookup v aa  ≡⟨ +-comm (lookup u aa) (lookup v aa) ⟩
+    lookup v aa + lookup u aa  ≡⟨ ≡.sym (⊕-lemma v u) ⟩
+    lookup (v ⊕ u) aa          ∎
+
+  ⊕-assoc : Algebra.Associative _≋_ _⊕_
+  ⊕-assoc x y z {a} = begin
+    lookup ((x ⊕ y) ⊕ z) a                  ≡⟨ ⊕-lemma (x ⊕ y) z ⟩
+    lookup (x ⊕ y) a + lookup z a           ≡⟨ cong (λ w → w + lookup z a) (⊕-lemma x y) ⟩
+    lookup x a + lookup y a + lookup z a    ≡⟨ +-assoc (lookup x a) (lookup y a) (lookup z a) ⟩
+    lookup x a + (lookup y a + lookup z a)  ≡⟨ cong (λ w → lookup x a + w) (≡.sym (⊕-lemma y z)) ⟩
+    lookup x a + lookup (y ⊕ z) a           ≡⟨ ≡.sym (⊕-lemma x (y ⊕ z)) ⟩
+    lookup (x ⊕ (y ⊕ z)) a                  ∎
+
+  ι-identity : Algebra.Identity _≋_ ι _⊕_
+  ι-identity .proj₁ tm {aa} = begin
+    lookup (ι ⊕ tm) aa          ≡⟨ ⊕-lemma ι tm ⟩
+    lookup ι aa + lookup tm aa  ≡⟨ cong (λ x → x + lookup tm aa) ι-lem3 ⟩
+    lookup tm aa                ∎
+  ι-identity .proj₂ tm {aa} = begin
+    lookup (tm ⊕ ι) aa          ≡⟨ ⊕-lemma tm ι ⟩
+    lookup tm aa + lookup ι aa  ≡⟨ cong (λ x → lookup tm aa + x) ι-lem3 ⟩
+    lookup tm aa + 0            ≡⟨ +-identityʳ (lookup tm aa) ⟩
+    lookup tm aa                ∎
+
+  open IsSemigroup
+  open IsMagma
+  isSemigrp : IsSemigroup _≋_ _⊕_
+  isSemigrp .isMagma .isEquivalence        = ≋-isEquivalence
+  isSemigrp .isMagma .∙-cong {u}{v}{x}{y}  = ⊕-cong {u}{v}{x}{y} -- (broken inference)
+  isSemigrp .assoc                         = ⊕-assoc
+
+  open IsMonoid
+  ≋-⊕-ι-isMonoid : IsMonoid _≋_ _⊕_ ι
+  ≋-⊕-ι-isMonoid .isSemigroup  = isSemigrp
+  ≋-⊕-ι-isMonoid .identity     = ι-identity
+\end{code}
+
+We are now in a position to define the commutative monoid.
+
+\begin{code}
+  open IsCommutativeMonoid
+  Vcm : CommutativeMonoid 0ℓ 0ℓ
+  Vcm .Carrier                        = AssetId ⇒ Quantity
+  Vcm ._≈_                            = _≋_
+  Vcm ._⋆_                            = _⊕_
+  Vcm .ε                              = ι
+  Vcm .isCommutativeMonoid .isMonoid  = ≋-⊕-ι-isMonoid
+  Vcm .isCommutativeMonoid .comm      = ⊕-comm
+
+  Value-TokenAlgebra :  (specialPolicy : PolicyId)
+                        (specialAsset : AssetName)
+                        (policies : AssetId ⇒ Quantity → ℙ PolicyId)
+                        (size : AssetId ⇒ Quantity → ℕ)
+                        {decTot : DecEq (AssetId ⇒ Quantity)}
```

Comment:
We may need to do something about this, since we can't ever supply this constructively. I think we need to require that the equivalence relation in `Vcm` is decidable, but I don't know how this affects everything else...

Reply:
Okay, thanks for pointing this out.  I'll have to think about how to resolve it.

Reply:
I think this is something we should leave for future consideration.